tests/test_inference_problem.py

```diff
@@ -0,0 +1,253 @@
+    def __init__(self):
+        super().__init__("Dummy")
+
+
+class NormalPrior:
+    def __init__(self, mean, sd):
+        self.mean = mean
+        self.sd = sd
+
+
+class ModelError:
+    def __call__(self, parameter_list):
+        raise NotImplementedError("Override this!")
+
+
+class LatentParameter:
+    def __init__(self):
+        self._affected_parameters = []
+        self.N = None
+        self.prior = None
+        self.start_idx = None
+
+    def _contains(self, key, name):
+        return (key, name) in self._affected_parameters
+
+    def add(self, key, name, N):
+        assert not self._contains(key, name)
+        if self.N is None:
+            self.N = N
+        else:
+            assert self.N == N
+
+        self._affected_parameters.append((key, name))
+
+    def __len__(self):
+        return len(self._affected_parameters)
+
+    def __getitem__(self, i):
+        return self._affected_parameters[i]
+
+    def __str__(self):
+        if self.N == 1:
+            idx_range = str(self.start_idx)
+        else:
+            idx_range = f"{self.start_idx}..{self.start_idx+N}"
+
+        return f"{idx_range:10} {self._affected_parameters} {self.prior}"
+
+
+class LatentParameterList:
+    def __init__(self):
+        self._all_parameter_lists = {}
+        self._latent = OrderedDict()
+        self._total_length = None
+
+    def add_parameter_list(self, parameter_list, key=None):
+        assert key not in self._all_parameter_lists
+        self._all_parameter_lists[key] = parameter_list
+
+    def add(self, latent_name, parameter_name, key=None):
+        assert key in self._all_parameter_lists
+        assert self._all_parameter_lists[key].has(parameter_name)
+
+        try:
+            N = len(self._all_parameter_lists[key][parameter_name])
+        except:
+            N = 1
+        
+        if latent_name not in self._latent:
+            self._latent[latent_name] = LatentParameter()
+
+        self._latent[latent_name].add(key, parameter_name, N)
+        self._update_idx()
+
+    def add_by_name(self, latent_name):
+        for key, prm_list in self._all_parameter_lists.items():
+            if prm_list.has(latent_name):
+                self.add(latent_name, latent_name, key)
+        
+
+    def _update_idx(self):
+        self._total_length = 0
+        for key, latent in self._latent.items():
+            latent.start_idx = self._total_length
+            self._total_length += latent.N or 0
+
+    def update(self, number_vector):
+        assert len(number_vector) == self._total_length
+        for l in self._latent.values():
+            latent_numbers = number_vector[l.start_idx : l.start_idx + l.N]
+            for (key, prm_name) in l:
+                self._all_parameter_lists[key][prm_name] = latent_numbers
+
+    def __str__(self):
+        s = ""
+        for key, latent in self._latent.items():
+            s += f"{key:10}: {latent} \n"
+        return s
+
+
+class Noise:
```

Comment:
We have to think about that and I was just toying around. I separated it because "LatentParameterList" (currently) only works on model parameters. 

I could not really come up with a nice way to include it..
~~~py
problem.noise.add("noise_name", sensor5, exp3_key)
# and later
problem.latent.add("noise_name", ???) 
~~~
Issue 1: They are _only_ coupled by the "noise_name" which the user has to drag around, if he wants to automate that. 
Issue 2: What to put in "???" because there is no ParameterList associated to the noise.

So one idea could be to have a 
~~~py
noise_prm = problem.noise.get_parameter_list()

noise_key = "SomeReallyDistinctKeyThatDoesNotInterfereWithPossibleUserKeys"
problem.latent.add_parameter_list(noise_prm, key=noise_key)
~~~

... actually quite OKish/goodish ;)

Reply:
I would have a 
```
problem.inference_prms.add('sigma_1', 0.1)
problem.inference_prms.add('YoungsModulus', stats.normal(30.000, 10.000))
```
the first variable is deterministic with a given value, the other is latent (due to giving a prior). These latent variables are the ones we loop over for the prior. As for the independent model errors I would suggest something like
```
inference_prms['sd1'] = 0.1 #deterministic
inference_prms['sd2'] = 10 #deterministic
inference_prms['sd3'] = stats.Gamma(s=0.1, c=3) #latent not sure about the actual distribution
inference_prms['l_corr_time] = 0.5 #correlation length
inference_prms['l_corr_space] = 0.5 

independent_model_errors.add_log_like(IndependentZeroMeanNormalNoiseModelError(key,sensor, 'sd1'))
independent_model_errors.add_log_like(TimeCorrelatedZeroMeanNormalNoiseModelError(key, sensor, time_steps, 'sd2', 'l_corr_time') #len(timesteps)=len(model_error(key,sensor)
independent_model_errors.add_log_like(SpaceTimeCorrelatedZeroMeanNormalNoiseModelError(key, sensor, time_steps, coordinates, 'sd3', 'l_corr_time', 'l_corr_space) #we will have to define the structure of a space_time_model error
```
And someone that wants to correlated two different sensortypes in space (e.g. because there is a laser that measures some points correlated in space and a stereo camera that measures also correlated in space, but both sensors are clearly not independent since they measure the same thing)
```
independent_model_errors.add_log_like(MyFancyModelError(key1, sensor1, key2, sensor2, coordinates1, coordinates2, 'sd1', 'sd2', 'l_corr_space', 'l_corr_laser_stereo')
```
 so in the likelihood, there will be a loop over the indenpent_model_errors each returning the contribution to the loglikelihood (a scalar value). In the special case of VB, you would only consider model errors of type IndependentZeroMeanNormalNoiseModelError (naming is certainly to be discussed)

Reply:
When it comes to e.g. model selection, I think it is advantageous for the user to automatically set deterministic parameter from some parameter files. So they should not explicitly be mentioned in the main file. The idea would be that we ask the model "what parameters to you need" and then look up a parameter file (or set them explicitly if needed) to provide those. The latent parameters, however, are in full control of the user and must be explicitly mentioned. And I would stick to this pattern for the noise models as well. So similar to the model_errors, the noise_models can tell the parameter it needs (initialized with `None`) and the user has to provide them. 

Your proposed idea is the other way round. You first define a parameter, e.g. "sd2" that you then pass to a noise model. The drawbacks I see:
* You accidentally write `independent_model_errors.add_log_like(TimeCorrelatedZeroMeanNormalNoiseModelError(key, sensor, time_steps, 'l_corr_time', 'sd2')` instead of your line. IMO hard to spot bug that "sd2" is now the fifth arg instead of fourth. That would also be against he "named parameter" idea because at this point, the meaning of the parameter is passed as the position in the argument list, not its name.
* I have to think about the focus (so mainly the naming) on "log_like". The user should IMO provide more general information that is later -- depending on the inference algorithm -- transformed into a loglike or a least-square vector. This information would just be: 
    * The output sensor2 in key1 is correlated in time with length 0.42s and has a zero mean noise with "sigma"=latent. 
    * The output of sensor1 in key1 is uncorrelated and has a systematic noise of a*t + b where a and b are latent.
    * The output of sensor3 in all keys (=model_errors) is correlated with 0.42s has known N(0., .6174)

The last example of combining individual coordinates is something we agreed not to allow. That came up when:
* one physical sensor has two noise terms (on bridge and off bridge). Here we agreed on splitting the sensor into two, each with his separate noise
* in a displacement field some displacements are already fixed via boundary conditions. So the "HugeDisplacementSensor" would be split accordingly.
* Same in your example: The measurement position of the stereo camera would also have to be split into individual sensors. 

But we may change that, sure.